### DISCUSSION TITLE
Add `entries` and `length` to `MemoryHistory`

### DISCUSSION THREAD

**Discussion by AlbertoSadoc** [remix-run/react-router#9853] [Proposals]
### What is the new or updated feature that you are suggesting?

The old `MemoryHistory` had those 2 properties that could be accessed in tests to check that the test navigated correctly through the paths provided

### Why should this feature be included?

Those properties existed in v5 and provided a quick and easy way to check all the paths. It will also make the upgrading easier, even if you mark the properties as deprecated to be removed later on

**Comment by joanvf** [comment]
Hi, is there any plan to address this? Our use case requires to know when to enable or disable our app's back and forward buttons. As of today, users click on the back or forward buttons and nothing happens if there is no history. In this case, these buttons should have been disabled. How can we achieve this?

Many thanks

**Reply by brophdawg11** [reply]
Generally the way to do this is to look for a `location.key` of `default` which is the value RR will set on the very initial load.  And navigations after that will have a unique key, and if the key is not default that is indicative of there being a location to navigate _back_ to.  Memory history is currently aligned to the API of `window.history` and I'm not sure we necessarily want to fork the API to differ between memory/browser/hash usages.

For testing purposes, it feels like you're trying to unit test internal RR behavior which is already well tested.  But if you need to do some of your own tests, instead of relying on internal entries - could you assert against the rendered output?  Even if that involves rendering the `location.key`?

**Comment by brophdawg11** [comment]
I'm going to convert this to a discussion so it can go through our new [Open Development](https://remix.run/blog/open-development) process.  Please upvote the new Proposal if you'd like to see this considered!

**Comment by 3alisaki** [comment]
i have same problem when there is no history the back or forward buttons should be disabled but i don't have any way to do that (the above way is only working for back button not for forward button).

**Comment by Donorlin** [comment]
Following on https://github.com/remix-run/history/pull/939

In our company, we are running SPA in the SSR (company historical reasons). This means the SPA state is lost whenever a site is reloaded. In version `history@4` we are able to save `entries` and `index` from the `history` object and pass it to `initialEntries` and `initialIndex` of the `MemoryRouter` to seed it back to its previous state (controlled behavior).

In `react-router@6` we are still able to access the `MemoryHistory` object with `UNSAFE_NavigationContext`, but not its entries. So we are not able to implement a controlled `MemoryRouter`.

sadly because of this, we are not able to upgrade to v6
this change wasnt even mentioned in changelog (not even a deprecation warning)

**Comment by Nantris** [comment]
I'd really appreciate if this feature was restored. We've been stuck and unable to upgrade for years now.
